Markdown support in the pinned items

kamilmielnik opened the pull request:
Closes https://github.com/metabase/metabase/issues/30381

I found the following issues while working on this:
- #31326
- #31327
- #31328

### Description

Adds markdown support in tooltips in pinned items.
- Headings and links are disabled in tooltips (to be consistent with changes in #29890)


### How to verify

1. Create a:
    - Dashboard
    - Model
    - Question
2. Add description to it:
    - markdown with short / long text
    - plain text with short / long text
3. Pin it to a dashboard
4. If it's a question, turn the visualization on and off
5. In the pinned item hover the text description (or description icon) to display the tooltip - it should show markdown without headings and links formatted
6. If there is a description block in the card, it should contain only the first description line with markdown formatting removed

To verify skeletons, check out the description tooltip when card is loading. To keep it in that state forever you can block requests to `api/card/*/query` URL in your developer tools.

### Demo

#### Before

https://github.com/metabase/metabase/assets/6830683/bf7ba71f-3eed-4a42-8309-8765ad09915b


#### After

https://github.com/metabase/metabase/assets/6830683/d501b58e-32a9-4e1b-a4cb-1f642dda9b70

kamilmielnik committed:
Allow ReactNode as tooltip in Ellipsified
- Consistently destructure all imports from react package

kamilmielnik committed:
Introduce getMarkdownFirstLine

kamilmielnik committed:
- Extract markdownToHtml
- Refactor getMarkdownFirstLine into getLeadingText

kamilmielnik committed:
Render description as markdown in StaticSkeleton

kamilmielnik committed:
Cover getLeadingText with tests

kamilmielnik committed:
Consolidate tests

kamilmielnik committed:
Extract test-utils

kamilmielnik committed:
Add initial markdownToHtml tests

kamilmielnik committed:
Add a test case for a mix of h1 and p

kamilmielnik committed:
Rename markdownToHtml to convertMarkdownToHtml

kamilmielnik committed:
Rename convertMarkdownToHtml to renderMarkdownToHtml

kamilmielnik committed:
Rename renderMarkdownToHtml to renderMarkdown

kamilmielnik committed:
Add a test case for alt text of an image
- e.g. Gmail does this and it's not the desired behavior

kamilmielnik committed:
Place the tooltip at the bottom (as per design)

kamilmielnik committed:
Add StaticSkeleton initial test suite

kamilmielnik committed:
Add test case for not showing description tooltip when description is fully rendered

kamilmielnik committed:
Rename renderMarkdown to parseMarkdown

kamilmielnik committed:
Format code

kamilmielnik committed:
Add an e2e test for markdown rendering in pinned models
- Introduce changeCardDescription helper in e2e-collection-helpers
- Move openRootCollection to e2e-collection-helpers

kamilmielnik committed:
Add an e2e test for markdown rendering in pinned native questions

kamilmielnik committed:
Remove redundant assertions

kamilmielnik committed:
Add an e2e test for markdown rendering in pinned dashboards

kamilmielnik committed:
Remove exact: false

kamilmielnik committed:
Move getLeadingText & parseMarkdown to metabase/lib/markdown

kamilmielnik committed:
Add markdown tooltip to PinnedItemCard

kamilmielnik committed:
Add missing import

kamilmielnik committed:
Add disallowHeading & unstyleLinks props to markdown rendered in tooltips

kamilmielnik committed:
Add test for markdown tooltip in question visualizations

kamilmielnik committed:
Support markdown formatting in Skeleton tooltip
- add an e2e test case for it
- rename changeCardDescription to changePinnedCardDescription

kamilmielnik committed:
Fix skeleton test

kamilmielnik committed:
Add an e2e test for model description tooltip on hover

kamilmielnik committed:
Consolidate pinned model description tooltip tests

kamilmielnik committed:
Wrap pinned dashboard test in describe

kamilmielnik committed:
Add an e2e test for dashboard description tooltip on hover

kamilmielnik committed:
Add an e2e test for pinned native question description tooltip on hover

kamilmielnik committed:
Add an e2e test for pinned model question description tooltip when ellpsis is and isn't applied

kamilmielnik committed:
Add an e2e test for pinned dashboard question description tooltip when ellipsis is and isn't applied

kamilmielnik committed:
Remove unit tests for StaticSkeleton as these cases are covered in e2e tests

kamilmielnik committed:
Merge branch 'master' into 30381-markdown-in-pinned-items

kamilmielnik committed:
Rename changeDashboardDescription to changePinnedDashboardDescription

ranquild commented on frontend/src/metabase/lib/markdown/parseMarkdown.tsx:
```diff
@@ -0,0 +1,16 @@
+import ReactDOMServer from "react-dom/server";
```
This is not something we'd want to use on the client

kamilmielnik replied:
> This is not something we'd want to use on the client

Why not?

BTW `frontend/src/metabase/static-viz/index.js` already uses it.

> `:first-line` selector seems promising https://caniuse.com/css-first-line

How would we use it?
Keep in mind that:
- [not all CSS properties](https://developer.mozilla.org/en-US/docs/Web/CSS/::first-line#allowable_properties) can be used with it.
- markdown may contain non-text nodes, e.g. images (see unit tests for it) - if the first node is not a text node `:first-line` won't help

> Instead of doing all the parsing, could you rely on CSS instead? 

I'm open for suggestions but I see nothing wrong with the current approach.

ranquild replied:
Static viz uses it on the server, not in browser

ranquild replied:
As for the explanation why I think it's a problematic approach. `react-dom/server` is a heavy dependency, it relies on the nodejs runtime. It works because we use webpack which polyfills used native nodejs modules, but it seems to be an overkill for a cosmetic change. 

If this task is something that cannot be achieved with a simpler approach, let's start a discussion with product people and see if there are alternatives.

ranquild replied:
@kamilmielnik tagging again just to make sure you get a notification - I think the new approach from the comment is good

kamilmielnik replied:
41e627ef35...d702b570d19da2a29f282dbc577d0bf361d4ad7a

ranquild reviewed (commented):
Added comments inline

ranquild reviewed (commented):
@kamilmielnik I think the API can be improved. From the consumer point of view, I want to render `Markdown` and say that I only want to show a single line. Wdyt about that? This way we could also consolidate all markdown related logic in the component folder, and make it more straightforward to use.

kamilmielnik committed:
Update parseMarkdown to use remark instead of react-dom/server
- We need to add remark-parse, remark-hype, and unified to package.json but these are not new - they are dependencies of react-markdown
- Tests will be updated in subsequent commits

kamilmielnik committed:
Remove rehype-filter from parseMarkdown

kamilmielnik committed:
Bring rehypeFilter back as it fixes typing errors in getLeadingText

kamilmielnik committed:
Remove unit tests for getLeadingText and parseMarkdown
- remark-parse used by parseMarkdown is an ESM without a minified
  commonjs-compliant bundle. We'd have to mock it to make it work
  but it'd defy the purpose of these tests. They're trivial enough with
  the usage of remark anyway.

kamilmielnik committed:
Remove unused test-utils

kamilmielnik committed:
Merge branch 'master' into 30381-markdown-in-pinned-items

kamilmielnik commented:
> I think the API can be improved. From the consumer point of view, I want to render Markdown and say that I only want to show a single line. Wdyt about that?

Sorry, I don't follow. Can you please provide some pseudocode to illustrate the idea?

> This way we could also consolidate all markdown related logic in the component folder, and make it more straightforward to use.

I'm not sure if this is a good idea:
- those markdown-related utility functions are not used by those components
- I haven't found a case in `metabase/core/components` where a utility function is exported
- I'd expect general utilities to live in general `utils`/`lib` etc. (think POLA)

ranquild commented on frontend/src/metabase/visualizations/components/skeletons/StaticSkeleton/StaticSkeleton.tsx:
```diff
@@ -44,7 +52,19 @@ const StaticSkeleton = ({
+            </Markdown>
+          }
+        >
+          {getLeadingText(defaultedDescription)}
```
@kamilmielnik I think that we parse markdown 2 times with the current API, could that be avoided?

kamilmielnik replied:
e120c406717db8478c41511ec55e67781c4eadbc

ranquild commented on frontend/src/metabase/lib/markdown/getLeadingText.ts:
```diff
@@ -0,0 +1,20 @@
+import { parseMarkdown } from "./parseMarkdown";
+import type { Node } from "./types";
+
+export const getLeadingText = (value: string): string => {
```
We don't create a separate module per (utility) function. In this case it would be a single `markdown.ts` module with all the functions in the folder. If you feel doing otherwise I'd recommend starting an RRC and getting approvals first before changing the approach.

kamilmielnik replied:
a6ffcfff32a790e45d017861fbfb4bab7e47fe6c

ranquild commented on frontend/src/metabase/lib/markdown/parseMarkdown.tsx:
```diff
@@ -0,0 +1,20 @@
+
+const REMARK_PLUGINS = [remarkGfm];
+
+export const parseMarkdown = (value: string, options: Options = {}) => {
```
I don't see tests for `markdown` functions

kamilmielnik replied:
> I don't see tests for markdown functions

When I try to test `parseMarkdown` which imports `remark-parse` (or any other `remark-*` lib) I get an error because it uses ESM import syntax.
`react-markdown` luckily comes with `min.js` which is [configured to be imported instead](https://github.com/metabase/metabase/blob/master/jest.unit.conf.json#L8) of default ESM.
`remark-*` libs come as pure ESM and if we import them directly into our code, tests fail.

@ranquild any suggestions for this?

kamilmielnik replied:
> any suggestions for this?

I had to enable transformations for these dependencies: 9236735f1b9c659dadaeff24c1d6136b5f0f9951
These are naturally required for unit tests of `PinnedItemCard` anyway.

kamilmielnik replied:
> I don't see tests for `markdown` functions

d2290e2056803e6f7495e83743b4fb5213bcc541 + fbcc2b9d0817738cbe2ca75649906d6fd2b00e1c

ranquild commented on e2e/test/scenarios/collections/collection-pinned-overview.cy.spec.js:
```diff
@@ -241,9 +265,244 @@ describe("scenarios > collection pinned items overview", () => {
-  cy.visit("/collection/root");
-  cy.wait("@getPinnedItems");
-};
+  describe("scenarios > collection pinned items overview > pinned model description tooltip", () => {
```
There are no backend calls or browser-specific logic (e.g. something rendering or layout-specific) that we test here, so I think we could (should?) test it in Jest in tests for the affected components.

ranquild replied:
There is a **draft** doc that clarifies the part about e2e. It's probably the best I have right now https://www.notion.so/metabase/Shift-the-engineering-quality-strategy-c75e948910654f348ec345c20b7044f6?pvs=4#ca158621ff3a4f92a9d4384fa053e1de

ranquild replied:
In my own words, I'd say that we add e2e tests for:
- any new shape or data coming from the BE
- any flows that involve interacting with the BE **and** processing the response, e.g. updating an entity or something similar
- cross-feature functionality, e.g. something that isn't scoped to a particular area and would require rendering the whole app

ranquild reviewed (commented):
Looks good, left some comments

kamilmielnik committed:
Merge branch 'master' into 30381-markdown-in-pinned-items

kamilmielnik committed:
Remove dead link to Source Sans Pro font
- it caused markdown check to fail during CI, this fixes it
- it seems that the font has been removed from Google Fonts
- users can still choose this font in Metabase settings
- removal confirmation: https://metaboat.slack.com/archives/C5XHN8GLW/p1686320576992299?thread_ts=1686319384.637909&cid=C5XHN8GLW

kamilmielnik committed:
Merge branch 'master' into 30381-markdown-in-pinned-items

kamilmielnik committed:
Merge markdown module into a single file

kamilmielnik committed:
Optimize getLeadingText

kamilmielnik committed:
Optimize getLeadingText

kamilmielnik committed:
Don't export types from markdown.ts
- and format code

kamilmielnik committed:
Configure transformIgnorePatterns to allow transforming ESM-based markdown-related dependencies

kamilmielnik committed:
Add unit tests for parseMarkdown

kamilmielnik committed:
Add unit tests for getLeadingText

kamilmielnik committed:
Remove redundant joins

kamilmielnik committed:
Add description-related unit tests to PinnedItemCard

kamilmielnik committed:
Add tests for ellipsis

kamilmielnik committed:
Add tests for description in StaticSkeleton

kamilmielnik committed:
Revert e2e tests

kamilmielnik committed:
Merge branch 'master' into 30381-markdown-in-pinned-items

kamilmielnik committed:
Merge branch 'master' into 30381-markdown-in-pinned-items

kamilmielnik committed:
Revert redundant change

kamilmielnik committed:
Add a SkeletonCaption test suite with test for markdown support in tooltip

kamilmielnik commented:
@ranquild I've addressed all comments. It's ready for re-review.

kamilmielnik committed:
Merge branch 'master' into 30381-markdown-in-pinned-items

kamilmielnik committed:
Make SkeletonCaption tooltip placement consistent with the one in StaticSkeleton

ranquild reviewed (commented):
Responded in Slack

kamilmielnik commented:
> Responded in Slack

The suggestion in Slack was [this prototype](https://github.com/metabase/metabase/pull/31573).

There are differences in how these 2 approaches will work:

1. This PR: only the first line of markdown will be rendered.
![image](https://github.com/metabase/metabase/assets/6830683/50cb3451-bf95-4b26-bf7b-807541faa9db)

2. Suggested approach: all markdown will be rendered.
![image](https://github.com/metabase/metabase/assets/6830683/5e814ef1-6bdd-4f79-8943-c6077ae736b0)

kamilmielnik commented:
@alxnddr Please see my [comment above](https://github.com/metabase/metabase/pull/31251#issuecomment-1592773212).

I believe this was Alexander's last remark/suggestion for this PR.
After thinking it through I don't think it's worth doing - the behavior would be affected negatively.
Also it does not feel like we'd gain much (we wouldn't need to add `metabase/lib/markdown.ts`) given the effort and aforementioned consequence.

alxnddr reviewed (commented):
@kamilmielnik I tried Alexander's approach and it worked correctly for me, and codewise looks more concise without adding new packages. How can I repro the issue you pointed out to? Is it solvable by `white-space: pre`?
<img width="687" alt="Screen Shot 2023-06-15 at 3 18 17 PM" src="https://github.com/metabase/metabase/assets/14301985/63cbe485-cd3a-4e34-a4e1-feb6145ba6cc">

kamilmielnik committed:
Merge branch 'master' into 30381-markdown-in-pinned-items

kamilmielnik committed:
Remove metabase/lib/markdown usage from PinnedItemCard

kamilmielnik committed:
Remove metabase/lib/markdown usage from StaticSkeleton

kamilmielnik committed:
Remove metabase/lib/markdown and related changes

kamilmielnik committed:
Update PinnedItemCard unit tests

kamilmielnik committed:
Add extra assertion to SkeletonCaption tests

kamilmielnik committed:
Update StaticSkeleton unit tests

kamilmielnik commented:
> Is it solvable by `white-space: pre`?

`pre` breaks ellipsis but `pre-line` does the job, thank you!
Ready for review @alxnddr.

alxnddr commented on frontend/src/metabase/dashboard/containers/DashboardHeader.jsx:
```diff
@@ -440,6 +440,7 @@ class DashboardHeader extends Component {
             <DashboardHeaderButton
               icon="info"
               isActive={isShowingDashboardInfoSidebar}
+              data-testid="dashboard-header-info-button"
```
this `data-testid` is unused and in general it is better to assign an `aria-label` to icon buttons

kamilmielnik replied:
575c7596f2215397f518010c9b05b84b32dd39b7

alxnddr commented on frontend/src/metabase/visualizations/components/skeletons/StaticSkeleton/StaticSkeleton.styled.tsx:
```diff
@@ -14,9 +15,15 @@ export const SkeletonTitle = styled(Ellipsified)`
 `;
 
-export const SkeletonDescription = styled(Ellipsified)`
+export const TruncatedMarkdown = styled(Markdown)`
```
nit: both truncated markdown components look very similar, it would be nice to reduce duplication if possible and makes sense

kamilmielnik replied:
Extracted `metabase/core/MarkdownPreview` component: f6106ffe4793c35d3e017a4fd6fbe4ce93546b4d + 7bbc68bf716654f19c6b76ba33865e7cdf1b0d69
Requesting re-review as this is the first component that I'm adding.

alxnddr reviewed (approved):
Looks great! Please check out a couple of minor comments before merging

kamilmielnik committed:
Merge branch 'master' into 30381-markdown-in-pinned-items

kamilmielnik committed:
Remove unused attribute

kamilmielnik committed:
Extract MarkdownPreview component

kamilmielnik committed:
Require children in MarkdownPreview

kamilmielnik committed:
Fix build

ranquild commented on frontend/src/metabase/collections/components/PinnedItemCard/PinnedItemCard.unit.spec.js:
```diff
@@ -110,4 +128,29 @@ describe("PinnedItemCard", () => {
     });
   });
+
+  describe("description", () => {
```
I think since this is now handled by `MarkdownPreview` component, we should only have maybe 1 integration test here with a very basic case, and move advanced tests to the component itself

kamilmielnik replied:
02656114e633871245785e77bf494b836b7bc4d6

kamilmielnik committed:
Merge branch 'master' into 30381-markdown-in-pinned-items

kamilmielnik committed:
Move detailed markdown tests into new, dedicated MarkdownPreview test suite

kamilmielnik committed:
Merge branch 'master' into 30381-markdown-in-pinned-items
